R/board_url.R

```diff
@@ -141,12 +198,130 @@ pin_store.pins_board_url <- function(board, name, paths, metadata,
+pin_version_delete.pins_board_url <- function(board, name, version, ...) {
+  abort_board_read_only("board_url")
+}
+
 #' @export
 write_board_manifest_yaml.pins_board_url <- function(board, manifest, ...) {
   abort_board_read_only("board_url")
 }
 
 # Helpers ------------------------------------------------------------------
+
+board_url_error_message <- c(
+  "{.var urls} must resolve to either:",
+  "*" = "an unnamed character scalar, i.e. a single URL",
+  "*" = "a named character vector",
+  "*" = "a named list, where all elements are character scalars or vectors"
+)
+
+get_url_format <- function(urls) {
+  if (is_scalar_character(urls) && !is_named(urls)) {
+    return ("pins_yaml")
+  }
+  if (is_list(urls) && is_named(urls) && all(map_lgl(urls, is_character))) {
+    return ("manifest_content")
+  }
+  if (is.character(urls) && is_named(urls)) {
+    return ("vector_of_urls")
+  }
+  cli::cli_abort(
+    board_url_error_message,
+    class = "pins_error_board_url_argument",
+    urls = urls
+  )
+}
+
+get_manifest <- function(url, call = rlang::caller_env()) {
+  # if ends with "/", look for manifest
+  if (grepl("/$", url)) {
+    url <- paste0(url, manifest_pin_yaml_filename)
+  }
+
+  # if request fails or returns with error code
+  tryCatch(
+    {
+      resp <- httr::GET(url)
+      httr::stop_for_status(resp)
+    },
+    error = function(e) {
+      cli::cli_abort(
+        message = c(
+          "Failed to access manifest file at {.url {url}}:",
+          " " = "{e$message}"
+        ),
+        class = "pins_error_board_url_request",
+        parent = e,
+        url = url,
+        call = call
+      )
+    }
+  )
+
+  # if file is not parsable
+  tryCatch(
+    {
+      text <- httr::content(resp, as = "text", encoding = "UTF-8")
+      manifest <- yaml::yaml.load(text)
+    },
+    error = function(e) {
+      cli::cli_abort(
+        message = c(
+          "Failed to parse manifest file at URL {.url {url}}:",
+          " " = "{e$message}",
+          "i" = "Manifest file must be text and parsable as YAML."
+        ),
+        class = "pins_error_board_url_parse",
+        parent = e,
+        resp = resp,
+        call = call
+      )
+    }
+  )
+
+  # prepend url-root to manifest entries
+  url_root <- url_dir(url)
+  manifest <- map(
+    manifest,
+    ~ map_chr(.x, ~ url_path(url_root, .x))
+  )
+
+  manifest
+}
+
+url_path <- function(url, ...) {
+  # return URL with path items appended
+  # - if last element in path ends with a slash,
+  #   ensure return URL ends with a slash
+
+  dots <- rlang::list2(...)
+  url_parsed <- httr::parse_url(url)
+  path_elems <- c(url_parsed$path, dots)
+
+  path_new <- rlang::exec(fs::path, !!!path_elems)
+  path_new <- end_with_slash(path_new)
+
+  url_parsed$path <- path_new
+
+  httr::build_url(url_parsed)
+}
+
+url_dir <- function(url) {
+  # return URL for the directory, end with slash
+  url_parsed <- httr::parse_url(url)
+  path <- fs::path_dir(url_parsed$path)
```

Comment:
Also @ijlyttle can you take a look at this example of using fs as well? Is it possible to suggest a test case that might illustrate why we need this?

Reply:
I'll take a closer look at alternatives. I know it's off-label use of fs, but my thought was that it's fs path-calculation functions being used for the path part of the URL, so I thought it would be OK.

I'll come back with more.